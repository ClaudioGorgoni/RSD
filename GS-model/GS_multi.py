import collections as collec
import sys
from multiprocess import Pool, cpu_count

import GS_sig_pi as gs
import numpy as np
from scipy.integrate import fixed_quad
from scipy.interpolate import splev


# The Legendre polynomials
def P_l(x, l):

	if l == 0:
		return 1.

	elif l == 2:
		return 0.5 * (3. * x * x - 1.)

	elif l == 4:
		return (35. * x ** 4. - 30. * x * x + 3.) / 8.

	else:
		sys.exit("Order of Legendre polynomial must be 0,2 or 4 !")



def xi_sigma_pi(s_par, s_per, fz, sigma_FoG, F1, F2):
	"""Return the GS integral value"""

	f11 = F1 * F2  # F'F''
	f20 = F1 * F1  # F'^2
	f02 = F2 * F2  # F''^2

	# Do the cubic interpolation of xi(r), v12(r), s12_par(r) and s12_perp(r)
	i_xi_r  = gs.interp_xi_r(   F1, F2, f20, f11, f02)
	i_v12   = gs.interp_v12(    F1, F2, f20, f11, f02)
	i_s12_p = gs.interp_s12_par(F1, F2, f20, f11, f02)
	i_s12_v = gs.interp_s12_per(F1, F2, f20, f11, f02)
	
	return gs.integrate_GS(s_par, s_per, i_xi_r, i_v12, i_s12_p, i_s12_v, fz, sigma_FoG)



def xi_smu(mu, s, alpha_p, alpha_v, fz, sigma_FoG, F1, F2):
	"""
	Return the GS integral value, given (mu, s)

	INPUT :
	-------
	mu        : (float) cosine angle between the LOS separation and the pair separation vectors
	s         : (float) scale length in the redshift space
	alpha_p   : (float) the parallel distortion parameter
	alpha_v   : (float) the perpendicular distortion parameter
	fz        : (float) the growth rate
	sigma_FoG : (float) finger of God velocity dispersion
	F1        : (float) 1st order Lagrangian bias
	F2        : (float) 2nd order Lagrangian bias

	OUTPUT :
	--------
	res       : (float) integrand of the GS model
	"""

	s_par = s * mu
	s_per = s * np.sqrt(1. - mu * mu)

	return xi_sigma_pi(alpha_p * s_par, alpha_v * s_per, fz, sigma_FoG, F1, F2)



def inner_integrand(mu, s, l, alpha_p, alpha_v, fz, sigma_FoG, F1, F2):
	"""
	Integrand of the multipole of order l
	---

	P_l(mu) is the Legendre polynomial of order l.
	l must be in [0,2,4]
	"""

	return xi_smu(mu, s, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) * P_l(mu, l) * (2.*l + 1.)
	
	

def vect_inner_integrand(mu, s, l, alpha_p, alpha_v, fz, sigma_FoG, F1, F2):
	
	foo = lambda x : inner_integrand(x, s, l, alpha_p, alpha_v, fz, sigma_FoG, F1, F2)
	return np.array(map(foo, np.array(mu)))



def mu_integral(args):
	"""
	Integrate "inner_integrand" along mu using quad

	INPUT :
	-------

	args  : (tuple) (s, l, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) 
			with s the comoving distance scale and l the multipole order
	"""

	return fixed_quad(vect_inner_integrand, 0., 1.,
					  args=(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7]))[0]



def xi_l(s, l, alpha_p, alpha_v, fz, sigma_FoG, F1, F2):
	"""
	Return the multipole prediction of order l for the CLPT+GS model

	"""	
	
	#if isinstance(s, (collec.Sequence, np.ndarray)):

	pool = Pool(cpu_count())

	tmp  = [ (ss, l, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) for ss in s]
		
	res  = np.array( pool.map(mu_integral, tmp) )

	#res = np.array( map(mu_integral, tmp) )
		
	pool.close()
	pool.join()

	#else:

	#	res = mu_integral((s, l, alpha_p, alpha_v, fz, sigma_FoG, F1, F2))

	return res


def xi_02(s, alpha_p, alpha_v, fz, sigma_FoG, F1, F2):
	"""
	Return the l=0,2 multipoles prediction for the CLPT+GS model

	"""	
	
	#if isinstance(s, (collec.Sequence, np.ndarray)):
	
	tmp =  [ (ss, 0, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) for ss in s]
	tmp += [ (ss, 2, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) for ss in s]
		
	pool = Pool(cpu_count())
			
	res  = np.array( pool.map( mu_integral, tmp) )
			
	pool.close()
	pool.join()
	
	#else:
	
	#	tmp  =  [ (s, 0, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) ]
	#	tmp  += [ (s, 2, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) ]
		
	#	pool = Pool()
			
	#	res  = np.array( pool.map( mu_integral, tmp) )
			
	#	pool.close()
	#	pool.join()
		
	return res


def xi_024(s, alpha_p, alpha_v, fz, sigma_FoG, F1, F2):
	"""
	Return the l=0,2,4 multipoles prediction for the CLPT+GS model

	"""

	#if isinstance(s, (collec.Sequence, np.ndarray)):

	tmp =  [(ss, 0, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) for ss in s]
	tmp += [(ss, 2, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) for ss in s]
	tmp += [(ss, 4, alpha_p, alpha_v, fz, sigma_FoG, F1, F2) for ss in s]

	pool = Pool(cpu_count())

	res = np.array( pool.map(mu_integral, tmp) )

	pool.close()
	pool.join()

	#else:

	#	tmp =  [(s, 0, alpha_p, alpha_v, fz, sigma_FoG, F1, F2)]
	#	tmp += [(s, 2, alpha_p, alpha_v, fz, sigma_FoG, F1, F2)]
	#	tmp += [(s, 4, alpha_p, alpha_v, fz, sigma_FoG, F1, F2)]

	#	pool = Pool()

	#	res = np.array( pool.map(mu_integral, tmp) )

	#	pool.close()
	#	pool.join()

	return res
