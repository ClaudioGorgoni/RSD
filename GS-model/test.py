#!/usr/bin/env python

import sys
import time

import matplotlib.pyplot as plt
import numpy as np
from GS_multi import xi_l, xi_02

r = np.linspace(10,150,16)

start = time.time()
xi0 = xi_l(r, 0, 0.98, 1.0, 0.74, 3.5, 0.95, -0.1)
xi2 = xi_l(r, 2, 0.98, 1.0, 0.74, 3.5, 0.95, -0.1)
print xi0, xi2
print time.time() - start


sys.exit()

s, xi0d = np.loadtxt('../DR12/Cuesta_2016_CMASSDR12_corrfunction_x0_prerecon.dat', usecols=(0,1,), unpack=True)
xi2d    = np.loadtxt('../DR12/Cuesta_2016_CMASSDR12_corrfunction_x2_prerecon.dat', usecols=(1,),   unpack=True)

cov = np.loadtxt('../DR12/Cuesta_2016_CMASSDR12_corrfunction_cov_x0x2_prerecon.dat')
err0 = np.sqrt(np.diag(cov[0:len(s),0:len(s)]))
err2 = np.sqrt(np.diag(cov[len(s):2*len(s),len(s):2*len(s)]))


plt.figure()
plt.plot(r,xi0*r*r,'r-',label='model')
plt.errorbar(s,xi0d*s*s,yerr=err0*s*s,fmt='bo',label='EZ mocks')
plt.xlabel(r'$r$')
plt.ylabel(r'$r^2\xi_0(r)$')
plt.xlim(10,150)
#plt.ylim(0,50)
plt.legend()
plt.grid()

plt.figure()
plt.plot(r,xi2*r*r,'r-',label='model')
plt.errorbar(s,xi2d*s*s,yerr=err2*s*s,fmt='bo',label='EZ mocks')
plt.xlabel(r'$r$')
plt.ylabel(r'$r^2\xi_2(r)$')
plt.xlim(10,150)
#plt.ylim(-70,0)
plt.legend()
plt.grid()

plt.show()



#s, xi0_d = np.loadtxt('../Correlation/output/SGC/58409.58576/ELG_v3_xi0.dat', usecols=(0,1), unpack=True)
#xi2_d    = np.loadtxt('../Correlation/output/SGC/58409.58576/ELG_v3_xi2.dat', usecols=(1),   unpack=True)

#cov  = np.loadtxt('../Covariance/EZ_mocks/SGC_cov_00.dat')
#err0 = np.sqrt(np.diag(cov))
#cov  = np.loadtxt('../Covariance/EZ_mocks/SGC_cov_22.dat')
#err2 = np.sqrt(np.diag(cov))

#rm, mm0, ms0 = np.loadtxt('../Covariance/EZ_mocks/SGC_mean_xi0.dat', usecols=(0,1,2), unpack=True)
#rm, mm2, ms2 = np.loadtxt('../Covariance/EZ_mocks/SGC_mean_xi2.dat', usecols=(0,1,2), unpack=True)

#rmq, mmq0, msq0 = np.loadtxt('../Covariance/qpm_mocks/SGC_mean_xi0.dat', usecols=(0,1,2), unpack=True)
#rmq, mmq2, msq2 = np.loadtxt('../Covariance/qpm_mocks/SGC_mean_xi2.dat', usecols=(0,1,2), unpack=True)

#plt.figure()
#plt.plot(r-1,xi0*r*r,'r-',label='fitted model')
#plt.plot(r-1,xi0_t*r*r,'g-',label='fiducial model')
#plt.errorbar(s,xi0_d*s*s,yerr=err0*s*s,fmt='bo',label='ELG data v3')
#plt.errorbar(rm,mm0*rm*rm,yerr=ms0*rm*rm,fmt='kx',label='EZ mocks')
#plt.errorbar(rmq+1,mmq0*rmq*rmq,yerr=msq0*rmq*rmq,fmt='m.',label='qpm mocks')
#plt.xlabel(r'$r$')
#plt.ylabel(r'$r^2\xi_0(r)$')
#plt.legend()
#plt.grid()
#plt.xlim(20,150)
#plt.ylim(-30,50)

#plt.figure()
#plt.plot(r-1,xi2*r,'r-',label='fitted model')
#plt.plot(r-1,xi2_t*r,'g-',label='fiducial model')
#plt.errorbar(s,xi2_d*s,yerr=err2*s,fmt='bo',label='ELG data v3')
#plt.errorbar(rm,mm2*rm,yerr=ms0*rm,fmt='kx',label='EZ mocks')
#plt.errorbar(rmq+1,mmq2*rmq,yerr=msq2*rmq,fmt='m.',label='qpm mocks')
#plt.xlabel(r'$r$')
#plt.ylabel(r'$r\xi_2(r)$')
#plt.legend()
#plt.grid()
#plt.xlim(20,150)
#plt.ylim(-2.,1)

#plt.show()



