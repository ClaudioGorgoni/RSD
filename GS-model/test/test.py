#!/usr/bin/env python2

import numpy as np
import emcee
import os
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import sys
import time
from getdist import plots, MCSamples

plt.rc('font',  family='serif')
plt.rc('text',  usetex=True)
plt.rc('xtick', labelsize=16) 
plt.rc('ytick', labelsize=16)

def gelman_rubin(chain):
    ssq = np.var(chain, axis=1, ddof=1)
    W   = np.mean(ssq, axis=0)
    Tb  = np.mean(chain, axis=1)
    Tbb = np.mean(Tb, axis=0)
    m   = chain.shape[0] * 1.0
    n   = chain.shape[1] * 1.0
    B   = n / (m - 1.) * np.sum((Tbb - Tb)**2., axis=0)
    var = (n - 1.) / n * W + 1. / n * B
    RR  = np.sqrt(var / W)
    return RR

# Choose the "true" parameters.
m_true = -0.9594
b_true = 4.294
f_true = 0.534

# Generate some synthetic data from the model.
N = 50
x = np.sort(10*np.random.rand(N))
yerr = 0.1+0.5*np.random.rand(N)
y = m_true*x+b_true
y += np.abs(f_true*y) * np.random.randn(N)
y += yerr * np.random.randn(N)

def lnlike(theta, x, y, yerr):
    m, b, lnf = theta
    model = m * x + b
    inv_sigma2 = 1.0/(yerr**2 + model**2*np.exp(2*lnf))
    return -0.5*(np.sum((y-model)**2*inv_sigma2 - np.log(inv_sigma2)))

import scipy.optimize as op
nll = lambda *args: -lnlike(*args)
result = op.minimize(nll, [m_true, b_true, np.log(f_true)], args=(x, y, yerr))
#m_ml, b_ml, lnf_ml = result["x"]

def lnprior(theta):
    m, b, lnf = theta
    if -5.0 < m < 0.5 and 0.0 < b < 10.0 and -10.0 < lnf < 1.0:
        return 0.0
    return -np.inf

def lnprob(theta, x, y, yerr):
    lp = lnprior(theta)
    if not np.isfinite(lp):
        return -np.inf
    return lp + lnlike(theta, x, y, yerr)

#-----------------------------------------------------------------


ndim, nwalkers = 3, 16
p0 = [result["x"] + 1e-4*np.random.randn(ndim) for i in range(nwalkers)]


sampler = emcee.EnsembleSampler(nwalkers, ndim, lnprob, args=(x, y, yerr))

pos,prob,state = sampler.run_mcmc(p0,100)
sampler.reset()
sampler.run_mcmc(pos, 1000)

labels = [r'$m$', r'$b$', r'$\ln(f)$']

# Burn the first 'mcmc_burn' steps of the walkers
samples = sampler.chain[:, :, :].reshape((-1, ndim))
chains  = sampler.chain[:, :, :]

for k in range(ndim):
	print gelman_rubin(chains[:,:,k]) - 1.
	
label = 'test'



for i in range(ndim):
	
	tmp = sampler.chain[:,:,i].T
	np.savetxt('chain_'+str(i+1)+'-'+str(label)+'.dat', tmp)
#---------------------------------------------------------------------------------


for i in range(ndim):
	
	tmp = sampler.chain[:,:,i].T
	
	plt.figure()
	plt.plot(tmp, '-', color='k', alpha=0.3)
	plt.xlabel(r'step number', size=16)
	plt.ylabel(labels[i], size=16)
	plt.savefig('chain_'+str(i+1)+'-'+str(label)+'.png', bbox_inches='tight', format='png')



samples[:, 2] = np.exp(samples[:, 2])
m_mcmc, b_mcmc, f_mcmc = map(lambda v: (v[1], v[2]-v[1], v[1]-v[0]),
                             zip(*np.percentile(samples, [16, 50, 84],
                                                axis=0)))

content  = ("m = %.3f (+%.3f / -%.3f)\n" % ( m_mcmc[0],  m_mcmc[1],  m_mcmc[2]))
content += ("b = %.3f (+%.3f / -%.3f)\n" % ( b_mcmc[0],  b_mcmc[1],  b_mcmc[2]))
content += ("f = %.3f (+%.3f / -%.3f)\n" % ( f_mcmc[0],  f_mcmc[1],  f_mcmc[2]))

print content
#mcmc_res.write(content)
#mcmc_res.close()

labels = ["m", "b", "f"]

sample1 = MCSamples(samples = samples, labels=labels)
g = plots.getSubplotPlotter()
g.triangle_plot(sample1, filled=True)
g.export('triangle_plot1.pdf')
g.export('triangle_plot1.png')

#plt.show()
