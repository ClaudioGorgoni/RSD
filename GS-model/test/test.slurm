#!/bin/bash

#SBATCH -p p4
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=4
#SBATCH --exclusive
#SBATCH -J test
#SBATCH -o test.stdout
#SBATCH -e test.stderr

module unload mpi4py

export OMP_NUM_THREADS=1
export NUM_THREADSPROCESSES=${SLURM_CPUS_PER_TASK}

./test.py
