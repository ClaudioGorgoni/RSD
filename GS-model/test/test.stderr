/astro/soft/common/gcc/5.4.0/python/2.7.15/lib/python2.7/site-packages/scipy/fftpack/basic.py:160: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  z[index] = x
